### DISCUSSION TITLE
How to use ACL in production

### DISCUSSION THREAD

**Discussion by pawelebe** [ory/keto#542] [Ideas]
Hey

I have some hard time figure out how I can integrate ACL and especially `keto` in real world application. Especially I am curious about consistency of data.

1. Where should I store group and user relations? Possible ideas:
* group, users and membership in relational db and just replicate it to keto?
* just store user and group metadata in db but membership store in keto?

If I store user membership in keto, how should I make sure it is persisted? Is there anything like transaction here? May I rely that data won't be lost?

2. I want to create UI for my permission system. I see that `expand` can show information like what are relations to given resource. But how you recommend to show that in UI? Is it expected that client parse somehow response ? So should I treat that as kind on NoSql where on client side I should store logic ?

3. What if I want get large amount of data from `keto`, or use `keto` information in some kind of join query in other part of the system. How you would recommend to do so, Examples may be:
* I store in `keto` what  users can current user see, then I want to show all of them. It may be thousand of them. So may I take all of them? Can I get for example 5th page without need to fetch all previously? Can I get count?
* what if I e.g. have `posts` in my relational db and want to see posts of all users for which current user has access? Should I  get all users ids from `keto` and then do query in my relational db? or maybe I should replicate  `can-see-post` relation in keto?

**Comment by zepatrik** [comment]
So Keto does require a database, which answers your questions around persistence and backups. Keto does not store any meta data about e.g. groups, that is something your application has to do. Membership however should only be stored through Keto.

For question 2 it really depends on your UI how you parse and display the data. There is a bit of info in the example: https://www.ory.sh/keto/docs/examples/olymp-file-sharing

If for 3 you exceed the capabilities of the List API, I would be happy to know more about your use case :wink: Generally, you cannot skip pages and there is a max page size.

> what if I e.g. have posts in my relational db and want to see posts of all users for which current user has access?

The exact use case would be interesting here, as there is no UI that can display all `posts` at the same time. You will always have a certain limit of how much stuff you want at once. If you plan on doing data analytics, it would probably make sense to directly use the database instead of Keto. 

Regarding transactions, the Change API does operate transactionaly, so you can either create all or none, delete some and create some or do nothing, ...
Maybe an interesting pointer you missed yet is https://www.ory.sh/keto/docs/concepts/api-overview

**Reply by pawelebe** [reply]
thanks @zepatrik for the blazing fast response!

Ok so lets assume that I store membership in `keto`, but still I can see some cases when I need this information in my primary db as well. For some kind of reporting/querying in my primary db. So I curious if it is ok and recommend replicating membership and store in both places? But then I am a little afraid of it getting out-of-sync. Is it a good approach or maybe I am barking on the wrong tree?

>If for 3 you exceed the capabilities of the List API, I would be happy to know more about your use case wink Generally, you cannot skip pages and there is a max page size.

Ok so let's say that I want to build user directory, where you can see only visible to you users. As I understand `keto` is perfect to check `can-i-see-that user`. However to get list of users there may be constraints regarding performance (i.e. getting lot of pages, especially when I want to see nth chunk of data). It may also get complicated when I want to have sorting or search for user list. Do you have any insight if ACL system is a good candidate of that?  Maybe then it should be kind of hybrid solution?

**Reply by zepatrik** [reply]
To get the info into your application, you can just ask Keto to get them. The List and Expand APIs are the way to go. As you said, expand will require more resources so only use it when required.

Good point about searching and sorting. This is indeed not possible. The idea is that you run the search in your application and filter the results based on Keto (remember, it has no meta data anyways). There are only rare cases where your application will have to search **all** objects. Most of the time e.g. a directory is searched or there is some other kind of context. Relating this to e.g. GitHub search, everything that is public should always be searched. Then you ask Keto to get all private repos a user has access to and search them as well. You would just not search all repos and filter out the results a user does not have access to. Maybe we can figure out some integrations with search engines like elastic that allow you to use them with Keto's permissions automatically.
Sorting should be solved similarly. You just sort the objects in the application instead of Keto. In some rare use cases you might need to add additional data to the objects in your application to figure out the context you are in.
As you suggested, it might also be feasible/necessary to go with a hybrid solution. We already have some thoughts around how that could work with e.g. OPA #318. Other hybrid approaches also make sense.

Regarding pagination, most modern APIs don't allow you to jump to a certain page. Again, there are only rare use cases where that is actually required. If you happen to find a specific one, I would be happy to discuss and find a solution for it.

**Reply by pawelebe** [reply]
Thanks for the reply @zepatrik . I was playing a bit trying to model a user directory example.

My problem is that `list` API showing only direct relations (I guess it means that subject-set is not expanded)

## Context
Let's assume our goal to set who may access user and rules are:
- user may be accessed by his group members
- user may be accessed by another group members, if there is group visibility link (implicit group)
- user may be accessed by some other user (explicit assignment)

## Use cases
### Model use cases (OK)

```
// set group membership
group:manager#member@(user:maureen)
group:manager#member@(user:laura)
group:director#member@(user:alice)

// set groups visibility 
group:manager#access@(group:manager#member)
group:director#access@(group:director#member)

// make group manager visible to group director
group:manager#access@(group:director#access)

// make users visible by their groups (1:1 membership)
user:maureen#access@(group:manager#access)
user:laura#access@(group:manager#access)
user:alice#access@(group:director#access)

// make explicit visibility for the user
user:maureen#access@(user:steven)
```

### Check if given user has access to another user (OK)
Works as a charm with `check` method

### Show who may access given user (?)
In order to do so I need to call `list` method several times

```
// to see which users (direct assignemtns) or groups may access him
keto relation-tuple get user --relation access --object maureen 

// his accessible by manager group, lets see who may see this group
keto relation-tuple get group --relation access --object manager 
// his groups was visible by director group, let's see who may access it
keto relation-tuple get group --relation access --object director 

// no more group-group access, so lets check members
// to see who are members of group manager
keto relation-tuple get group --relation member --object manager

// to see who are members of group director
keto relation-tuple get group --relation member --object director
```

So seems like to get this information i need to make `(1 + 2*NUMBER_OF_DIRECT_GROUPS + 2*NUMBER_OF_IMPLICIT_GROUPS)` requests.


Alternatively, I may just call `expand` (but not sure if it is the correct approach). This works fine for 10k users, but is very slow for 100k users

```
keto expand access user maureen                              
∪ user:maureen#access
├─ ∪ group:manager#access
│  ├─ ☘ group:director️
│  ├─ ∪ group:manager#member
│  │  ├─ ☘ user:laura️
│  │  ├─ ☘ user:maureen️
├─ ☘ user:steven️

```

###  Show what can given user see
Here I can also do several calls to `list` endpoint

```
// Get direct assignements
keto relation-tuple get user --relation access --subject user:alice

// I know that user may gain visibility through groups so lets check membership
keto relation-tuple get group --relation member --subject user:alice

// he is a member of director group so let see who may access that group
keto relation-tuple get group --relation access --subject group:director

// group director was visible to group manager so lets see who may access that
keto relation-tuple get group --relation access --subject group:manager

// no more group-group access, so lets see which users his groups may access
// let see what director groups may access
keto relation-tuple get user --relation access --subject group:director#access

// let see what manager groups may access
keto relation-tuple get user --relation access --subject group:manager#access

```
So seems like to get this information i need to make `(1 + 2*NUMBER_OF_DIRECT_GROUPS + 2*NUMBER_OF_IMPLICIT_GROUPS)` requests.

Moreover seems like this is the only possibility as expand won't work in this use case

## Conclusion

1. Is my approcach correct?
1. If there is lot of relations between groups or user is member of many groups, then seem that there will be a lot of requests needed.

**Reply by zepatrik** [reply]
> ### Show who may access given user

That part seems perfectly fine to me, both approaches are valid. It depends a bit on what exactly you need. One problem with expand is that it is not paginating the tree, but instead fetches all (in your case 100k) objects. Adding pagination there (although I am not yet sure how) should solve the problem. The bottleneck right now is most likely the database, if your system has enough memory to hold all the data.
If the use case is kind of like an audit feature where you just want to display a list, expand would be better suited. If you want to do some deeper analysis you might want to use the list API instead to follow your strategy of traversing the data.

> ### Show what can given user see

For that part you are right, the list API is the only way to solve this. In theory we can add another endpoint that covers this use case, as it is quite common. Basically, what expand does is traverse [the graph of relations](https://www.ory.sh/keto/docs/concepts/graph-of-relations) and output everything as a operation tree. This traversal can also be done in the other direction, basically what you have done. I am not quite sure if it will be very performant in both ways.
Generally, what you want is to control the basic shape of the graph of relations a bit. That way you can achieve much better performance. I have not written a lot about that in the docs yet, but there will be some sections explaining it in more detail with best-practices to follow. https://www.ory.sh/keto/docs/performance is scratching on the surface only :wink: 

### One key takeaway point is:

> I know that user may gain visibility through groups so lets check membership

You will usually know a lot about the structure of your graph. Because of that, you can use the list API quite effectively, and until there really comes up the need for an inverse expand, I would like to not add it. We really have to gather more real-world use cases and performance issues to start optimizing this. But be assured that there will be a solution, and if it is only that we improve the list API to allow getting all the results you described in a single call.

> f there is lot of relations between groups or user is member of many groups, then seem that there will be a lot of requests needed.

That part is true, it is more of an edge case IMO. Again. with an improved/new API this should not be a concern, especially when you throw in #312

**Reply by emayssat** [reply]
(1) I am not designing the product, that being said there is a difference between a user and a user entry in a directory. To me those are 2 different objects. 
(2) How you organize your ACLs (i.e associated mental model) is important to be able to optimize the queries. For what I understand queries should be optimized so that as soon as an answer can be given, the query does not have to be complete. For example, if you say a user can view an entry if she is in group1, group2, group3, ... OR groupN, then if she is found out to be in group1, you don't need to query for the other memberships (or wait for their results if queries have been sent). That level of optimization is not implemented in keto though, but should be for max performance. The idea remains that the ACLs should be organized/structured based on the expected type of queries. Those are database queries after all.
Again just my view.

**Reply by zepatrik** [reply]
> there is a difference between a user and a user entry in a directory

Right, thanks for pointing that out. You should be aware of that, but if you use the right data in the API calls you will be fine. In fact, in this example every user is once represented in the `user` namespace by it's username, and also as a subject by `user:{USERNAME}`. The string representation is the same, but internally it gets parsed to a namespace and object in the first case, but to a subject including the colon in the second case. If you want to prefix your users with some string to avoid clashes, that is a fine approach as well.

> That level of optimization is not implemented in keto though, but should be for max performance.

This is partially implemented, at least in the check engine.
https://github.com/ory/keto/blob/bad033b77aeed01222dcb4dd62d2d72f42a79aaf/internal/check/engine.go#L39-L42
https://github.com/ory/keto/blob/bad033b77aeed01222dcb4dd62d2d72f42a79aaf/internal/check/engine.go#L54-L56
More optimizations will be added with intersection and exclusion, as appropriate. Also, (local) caching within Keto will improve the performance regarding this.
